buildEnv: support fixed-point arguments

ShamrockLee opened the pull request:
1. This PR restructure `buildEnv` with `lib.extendMkDerivation`, `stdenvNoCC.mkDerivation`, and `buildCommand`. `buildEnv` can now take fixed-point arguments (`finalAttrs: { }`).
2. ~We still rely on the custom overrider `<env-pkg>.override`, as fixing the `__structuredAttrs =  true` support might change the  `<pkg>.overrideAttrs` interface.~
   PR #434815 proves that the argument overriding is mostly the same after adopting structured attributes. Let's deprecate the custom overrider `<env-pkg>.override` right away, but keep it until tree-wide transition to `<pkg>.overrideAttrs`.
4. This PR introduces the `derivationArgs` argument for `buildEnv` so that we won't have to exhaustively list all the `stdenv.mkDerivation` arguments users might wish to pass.
5. If tested, this PR minus the release note can be backported.

This PR depends on:
- #458529

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [x] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ShamrockLee committed:
buildEnv: construct with extendMkDerivation and stdenvNoCC.mkDerivation

ShamrockLee committed:
buildEnv: reference attributes from finalAttrs

ShamrockLee committed:
rl-2511.section.md: document buildEnv's supporting finalAttrs and <pkg>.overridAttrs

ShamrockLee committed:
buildEnv: take derivationArgs

ShamrockLee commented:
@ofborg build nixpkgs-vet

ShamrockLee commented:
I don't know why the CI complains that `nixpkgs-vet` fails to build. It builds locally.

philiptaron commented on pkgs/build-support/buildenv/default.nix:
```diff
@@ -13,114 +13,141 @@ let
+    extendDrvArgs =
+      finalAttrs:
+      {
+
```
Nit: extra newline.

philiptaron reviewed (commented):
My sense is that the rebuilds come about because of the switch from `runCommand` to `stdenvNoCC.mkDerivation`. Is that right?

The order of the merge operations (derivationArgs first, then applying the functions arguments) gives me some heartburn. It's always odd to have something set in derivationArgs which then doesn't take effect. It's sort of a bug factory. The other way is also perilous however, as certain derivationArgs are actually invalid for this build helper.

My preferred way to handle this is to use something like intersectAttrs to forbid a few things in derivationAttrs that would malform the resulting derivation. See `replaceVarsWith` for an example of that. Thanks to @wolfgangwalther for the pattern.

philiptaron commented:
I'm also interested in how we deprecate and remove the `lib.makeOverrideable` call.

ShamrockLee commented:
> Taking a look. Is there a version of these that doesn't rebuild everything?

It might be doable by excluding some arguments, but that contradicts the purpose of this PR.

wolfgangwalther reviewed (commented):
> We can support structured attributes while keeping `<pkg>.overrideAttrs` unchanged by refactoring `buildenv/builder.pl` to get variables from the JSON file whose path is inside the environment variable [`NIX_ATTRS_JSON_FILE`](https://hydra.nixos.org/build/305054836/download/1/manual/store/derivation/index.html#structured-attrs) if such environment variable is available and nonempty, and falling back to the old behaviour of getting from environment variables of the same names.

Sounds like the right approach. But since you're mentioning `overrideAttrs` explicitly, it sounds like there is *another* way of fixing structuredAttrs support which would not keep `overrideAttrs` the same? Which way would that be?

ShamrockLee commented on pkgs/build-support/buildenv/default.nix:
```diff
@@ -13,114 +13,142 @@ let
+    names: attrs: lib.getAttrs (lib.intersectLists names (lib.attrNames attrs)) attrs;
 in
 
+# We currently still rely on this custom overrider `<env-pkg>.override`,
+# as fixing the structured attribute support would change the `<pkg>.overrideAttrs` interface.
```
The approaches I considered includes:

1. Moving under `env` the variables to pass to the builder. This way we don't need to change the Perl build script.

2. Enforce `__structuredAttrs = true` for all constructed packages and pass everythong consistently via `NIX_ATTRS_JSON_FILE`. This way we avoid the manual serialization of `[ { paths = ; priority = ; } ]` completely.

ShamrockLee replied:
> This would not change the overrideAttrs interface either, right?

Yes, it won't change the overriding interface. (I personally feel like replacing `env.pkgs` with a structual `env.pathSpecs` and make it part of the public interface, so that people could add paths with custom priority.)

Nevertheless, it's a tree-wide backward-incompatible change whose impact spreads across several language- and framework-specific toolchain. Python's `python3Packagrs.buildEnv` is currently a `pkgs.buildEnv`-constructed package, and `python3.withPackabes` is implemented by overriding `python3Packages.buildEnv`. In addition, it provides `postBuild` to allow appending custom commands. In the worst case scenario, we'll have to test dependent packages and frameworks one by one and do our best to inform users about the breakage.

ShamrockLee replied:
> Enforcing structured attrs

I mean this one. we cannot control what people would put inside `postBuild` and which setup hooks they bring in via `nativeBuildInputs`.

ShamrockLee replied:
Still, I do agree that structured attributes are the right way to go. I'm just uncertain if we could break things without providing a backward-compatible workaround.

ShamrockLee replied:
It just came to me that we usually don't provide bavkward compatibility for `pre`- or `post`- stuff. Maybe we could try it on the `master` branch, whose users are expected to tolerate breakage, and fix things that break.

ShamrockLee replied:
> Since we set `buildCommand`, the phases will not be run. I think this means that setup hooks won't be run at all.
> 
> So we just need to deal with `postBuild`, which is manually ran in the buildCommand.
> 
> > So the breaking change would be, because the structured attrs file is not additionally loaded as bash variables?
> 
> I think this would actually already happen through `stdenv/setup.sh`.
> 
> So I don't see the breaking part, yet.

ah, i see.

I thought that the user might pass some random attributes and expect them to be passed to `buildCommand` as an environment variable, but that is not possible without the `derivationArgs` introduced in this PR or `overrideAttrs`. Since both are not currently supported, there shouldn't be ba kward incompatibility.

ShamrockLee replied:
As for the setup hooks that still fail to support structual attrs, let's consider such failure their bugs.

philiptaron replied:
> As for the setup hooks that still fail to support structural attrs, let's consider such failures to be their bugs.

Quite so. I agree with both you and Wolfgang this is the direction the PR ought to go.

ShamrockLee commented on pkgs/build-support/buildenv/default.nix:
```diff
@@ -13,114 +13,142 @@ let
-        extraPathsFrom = lib.optional includeClosures (writeClosure pathsForClosure);
-        preferLocalBuild = true;
-        allowSubstitutes = false;
+        extraPathsFrom = lib.optional finalAttrs.includeClosures (writeClosure pathsForClosure);
```
I noticed that `extraPathsFrom` is treated as a single file in `buildenv/builder.pl`. We should correct it and make it a string instead of a list.

ShamrockLee replied:
Applied.

ShamrockLee committed:
buildEnv: specify extraPathsFrom a string as expected by the builder

Specify `extraPathsFrom` as a string using `lib.optionalString` instead of a singleton list using `lib.optional` to align with the current behavior of the Perl build script.

ShamrockLee commented:
> > Taking a look. Is there a version of these that doesn't rebuild everything?
> 
> It might be doable by excluding some arguments, but excluding `paths` from the passing arguments makes it impossible to get its final state from `finalAttrs` and contradicts the purpose of this PR.

It came to me that we must not pass `paths` directly, as the chosen outputs does not necessarily contain the first output (whose `outPath` will be the string-interpolation result). Now `paths` is passed as `passthru.paths`.

I also found that some packages doesn't handle `passthru` careful enough when overriding, causing its attributes to be wiped out. Let's fall back to the original `paths` for now, and add a warning for this in subsequent PRs to prevent backward incompatibility.

ShamrockLee committed:
texlive: buildTeXEnv: use buildEnv with derivationArgs instead of custom buildEnv'

ShamrockLee committed:
texlive: buildTeXEnv: format expression

ShamrockLee committed:
doc: buildenv.section.md: init

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 432957 --package python3.buildEnv --package lua.buildEnv --package octave.buildEnv --package texlive.schemes`
Commit: `96447557740c410d0c1077f4e0043e5d169dbec3`

---
### `x86_64-linux`
<details>
  <summary>:x: 31 packages failed to build:</summary>
  <ul>
    <li>octave.buildEnv</li>
    <li>texlive.schemes.texliveBasic</li>
    <li>texlive.schemes.texliveBasic.info (texlive.schemes.texliveBasic.info.info, texlive.schemes.texliveBasic.info.man)</li>
    <li>texlive.schemes.texliveBasic.man (texlive.schemes.texliveBasic.man.info, texlive.schemes.texliveBasic.man.man)</li>
    <li>texlive.schemes.texliveBookPub</li>
    <li>texlive.schemes.texliveBookPub.info (texlive.schemes.texliveBookPub.info.info, texlive.schemes.texliveBookPub.info.man)</li>
    <li>texlive.schemes.texliveBookPub.man (texlive.schemes.texliveBookPub.man.info, texlive.schemes.texliveBookPub.man.man)</li>
    <li>texlive.schemes.texliveConTeXt</li>
    <li>texlive.schemes.texliveConTeXt.info (texlive.schemes.texliveConTeXt.info.info, texlive.schemes.texliveConTeXt.info.man)</li>
    <li>texlive.schemes.texliveConTeXt.man (texlive.schemes.texliveConTeXt.man.info, texlive.schemes.texliveConTeXt.man.man)</li>
    <li>texlive.schemes.texliveFull</li>
    <li>texlive.schemes.texliveFull.info (texlive.schemes.texliveFull.info.info, texlive.schemes.texliveFull.info.man)</li>
    <li>texlive.schemes.texliveFull.man (texlive.schemes.texliveFull.man.info, texlive.schemes.texliveFull.man.man)</li>
    <li>texlive.schemes.texliveGUST</li>
    <li>texlive.schemes.texliveGUST.info (texlive.schemes.texliveGUST.info.info, texlive.schemes.texliveGUST.info.man)</li>
    <li>texlive.schemes.texliveGUST.man (texlive.schemes.texliveGUST.man.info, texlive.schemes.texliveGUST.man.man)</li>
    <li>texlive.schemes.texliveInfraOnly</li>
    <li>texlive.schemes.texliveInfraOnly.info (texlive.schemes.texliveInfraOnly.info.info, texlive.schemes.texliveInfraOnly.info.man)</li>
    <li>texlive.schemes.texliveInfraOnly.man (texlive.schemes.texliveInfraOnly.man.info, texlive.schemes.texliveInfraOnly.man.man)</li>
    <li>texlive.schemes.texliveMedium</li>
    <li>texlive.schemes.texliveMedium.info (texlive.schemes.texliveMedium.info.info, texlive.schemes.texliveMedium.info.man)</li>
    <li>texlive.schemes.texliveMedium.man (texlive.schemes.texliveMedium.man.info, texlive.schemes.texliveMedium.man.man)</li>
    <li>texlive.schemes.texliveMinimal</li>
    <li>texlive.schemes.texliveMinimal.info (texlive.schemes.texliveMinimal.info.info, texlive.schemes.texliveMinimal.info.man)</li>
    <li>texlive.schemes.texliveMinimal.man (texlive.schemes.texliveMinimal.man.info, texlive.schemes.texliveMinimal.man.man)</li>
    <li>texlive.schemes.texliveSmall</li>
    <li>texlive.schemes.texliveSmall.info (texlive.schemes.texliveSmall.info.info, texlive.schemes.texliveSmall.info.man)</li>
    <li>texlive.schemes.texliveSmall.man (texlive.schemes.texliveSmall.man.info, texlive.schemes.texliveSmall.man.man)</li>
    <li>texlive.schemes.texliveTeTeX</li>
    <li>texlive.schemes.texliveTeTeX.info (texlive.schemes.texliveTeTeX.info.info, texlive.schemes.texliveTeTeX.info.man)</li>
    <li>texlive.schemes.texliveTeTeX.man (texlive.schemes.texliveTeTeX.man.info, texlive.schemes.texliveTeTeX.man.man)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>lua.buildEnv</li>
    <li>python3.buildEnv</li>
  </ul>
</details>

Ch1keen commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 432957 --package ruby.withPackages --package php.buildEnv --package perl.buildEnv`
Commit: `96447557740c410d0c1077f4e0043e5d169dbec3`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>perl.buildEnv</li>
  </ul>
</details>

Ch1keen commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 432957 --package ruby.withPackages --package php.buildEnv`
Commit: `96447557740c410d0c1077f4e0043e5d169dbec3`

---
### `x86_64-linux`

Ch1keen commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 432957 --package ruby.withPackages --package php.buildEnv`
Commit: `96447557740c410d0c1077f4e0043e5d169dbec3`

---
### `x86_64-linux`

Ch1keen commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 432957 --package keepass`
Commit: `96447557740c410d0c1077f4e0043e5d169dbec3`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>keepass</li>
  </ul>
</details>

ShamrockLee commented:
Rebased on top of `master` to resolve merge conflicts. (Fortunately, no manual intervention is needed.)

philiptaron reviewed (changes requested):
Requesting changes for:

1. The questions I have about `pname`, `version`, and `name`.
2. The situation with `derivationArgs` priority and overlap.

Beyond that, it would be so much easier to ✅ this PR with tests. There are many, many subtle scenarios and there are no tests in the tree that I can see. I find myself having a crisis of confidence in trying to work up the courage to merge this, even though I very much agree with its motivation. If there were a prior PR with tests, it would allow this PR to show that it changed the implementation but did _not_ change the produced artifacts.

I have a bunch of other feedback about documentation, which I don't think is blocking, but as long as we're writing things, can be sorted out.

ShamrockLee commented on pkgs/build-support/buildenv/default.nix:
```diff
@@ -13,114 +13,157 @@ let
+        meta ? null,
+      }@args:
+      let
+        compatArgs = getOptionalAttrs [
+          "nativeBuildInputs"
+          "buildInputs"
+          "meta"
+        ] args;
```
I hope to eventually place all of them into `derivationArgs`, so I choose to let the `derivtaionArgs`-specified meta win.

Current implementation of `writeTextFile` merges the two but prioritize sub-attributes from `derivationArgs`. Maybe we could adopt it.

ShamrockLee replied:
That makes sense. Both `meta` and `passthru` are special and not passing down to `builtins.derivation`. For that we could merge them, while giving prioritizing the ones specified directly.

Not sure about `postBuild`, though. It behaves quite different from the `runHook postBuild` in the generic `buildPhase()`. I would say that the current `eval "$postBuild"` is `buildEnv`-specific.

ShamrockLee replied:
BTW, we could ping people in PR #280592 for inputs about the design decision of `writeTextFile`'s `derivationArgs`.

@9999years, what do you think about the `passthru` and `meta` priority when introducing `derivationArgs`?

ShamrockLee commented on doc/build-helpers/special/buildenv.section.md:
```diff
@@ -0,0 +1,65 @@
+    and only if not, throw a collision error.
+
+-   `manifest` (default: `""`):
+    The manifest file (if any).
```
I guess it would be some structured plain-text file placed at the project root to define a package, something like `manifest.json`, `package.json` and `pyproject.toml`.

Frankly speaking, I don't think those are special enough to occupy a dedicated `buildEnv` argument. Does it have some even more important role to play in, e.g., the module system profile definition?

ShamrockLee commented on pkgs/build-support/buildenv/default.nix:
```diff
@@ -13,114 +13,157 @@ let
+            paths
+            ;
+        }
+        // derivationArgs.passthru or args.passthru or { };
```
Addressed.

ShamrockLee commented on doc/build-helpers/special/buildenv.section.md:
```diff
@@ -0,0 +1,65 @@
+# buildEnv {#sec-buildEnv}
+
+`buildEnv` constructs a derivation containing directories and symbolic links,
+which resembles the profile layout where a list of derivations or store paths are installed.
```
We could also consider formalizing the term *profile*, which is "a directory of symlinks to files in the Nix store." 

- https://hydra.nixos.org/build/310113865/download/1/manual/command-ref/files/profiles.html

ShamrockLee commented:
Rebase to resolve release note merge conflicts. Hope that we could find a sensible way to generate our release note without those silly merge conflicts in the future.

ShamrockLee commented:
The above rebase is also for rebasing PR #434815 after #451871 merged.

ShamrockLee commented:
> > 3. The situation with `derivationArgs` priority and overlap.
> 
> I can't find that discussion thread 👀

Here it is: https://github.com/NixOS/nixpkgs/pull/432957#discussion_r2410599625

wolfgangwalther commented:
> The error is caused by a package `ronin` having `meta.identifiers`, which is unrelated to this PR.

Why do you believe this is not related? I don't see it anywhere else.

ShamrockLee commented:
build-tex-env.nix changes dropped, and the eval tests stay red.

I'm going to drop the `meta` related changes in `buildEnv`.

wolfgangwalther commented:
You can run eval locally as well, just do `nix-build ci -A eval.singleSystem`.

Also you should be able to just try to eval `ronin.meta` or so (with checkMeta enabled I assume) to get the same warning?

RossSmyth reviewed (dismissed):
Very nice PR, would be cool to see it.

philiptaron committed:
buildEnv: fix passthru merging

Merge derivationArgs.passthru into the passthru chain so that passthru attributes passed via derivationArgs are not silently dropped.

philiptaron committed:
doc: buildenv.section.md: fix argument name, default, and grammar

- includeClosure -> includeClosures (matching the actual argument name)
- extraPrefix default: [ ] -> "" (matching the actual default)
- Fix broken sentence about ignoreSingleFileOutputs

philiptaron committed:
tests.buildenv: init test harness

Add an eval-time test harness for buildEnv following the pattern established by `tests.overriding`.
Uses `lib.runTests` to compare expr/expected pairs, and the buildCommand reports pass/fail.

No tests yet; subsequent commits add them individually.

philiptaron committed:
tests.buildenv: test name/pname/version handling

Verify that:
- `name` argument is used directly as the derivation name
- `pname` + `version` composes into the derivation name
- Omitting all three triggers the buildEnv-specific assertion

The assertion test uses `builtins.tryEval` + `builtins.seq` to force evaluation of .drvPath and confirm it throws.

philiptaron committed:
tests.buildenv: test passthru.paths injection

buildEnv stashes `paths` into `passthru.paths` (rather than passing it directly as a derivation attribute) to prevent unexpected string context pollution. Verify that:
- paths are accessible via passthru.paths
- passthru.paths can be overridden with overrideAttrs

philiptaron committed:
tests.buildenv: test finalAttrs self-reference

This is the core new capability of the PR: `buildEnv` now accepts a function `(finalAttrs: { ... })` enabling self-referencing attributes.

Verify that finalAttrs.name is accessible from within the argument function.

philiptaron committed:
tests.buildenv: test overrideAttrs

Verify that overrideAttrs on a buildEnv result:

- Can modify passthru attributes
- Preserves the derivation name
- Produces a different derivation when a build-affecting attribute (postBuild) is changed

philiptaron committed:
tests.buildenv: test passthru merging

buildEnv merges passthru from three sources:
  `{ inherit paths; } // derivationArgs.passthru // passthru`

Verify that:
- Auto-injected paths are present in passthru
- derivationArgs.passthru attributes are preserved
- Direct passthru attributes are preserved
- Direct passthru takes precedence over derivationArgs.passthru

philiptaron committed:
tests.buildenv: test derivationArgs forwarding

Verify that:
- derivationArgs attributes reach the underlying mkDerivation (tested via allowSubstitutes overriding the default false)
- The backward-compat layer for top-level nativeBuildInputs still works (these are forwarded through compatArgs)

philiptaron committed:
tests.buildenv: add build tests for output verification

Add derivations that actually build a buildEnv and verify the output tree.
Exposed as passthru.buildTests so they can be built individually or all at once.

- basic-symlinking: single package produces correct symlink structure
- pathsToLink: only specified subdirectories appear in output
- extraPrefix: output is rooted under the specified prefix
- postBuild: shell commands execute after the symlink tree
- ignoreCollisions: duplicate paths succeed when ignoreCollisions=true

philiptaron reviewed (approved):
I added a bunch of tests. Please take a look.

philiptaron commented on r-ryantm-500plus-rebuild-prs.md:
That is my mistake.

philiptaron replied:
I fixed this.

philiptaron commented on pkgs/development/python-modules/pybamboohr/default.nix:
Yeah, this is my mistake… I am occasionally a noob with git worktrees, and this is me reusing one that I shouldn't have.

philiptaron committed:
tests.buildenv: assert that structuredAttrs is broken

buildEnv's builder.pl reads all inputs from environment variables,
which is incompatible with structuredAttrs (values go into a JSON file
instead). Document this known failure as a test.

ShamrockLee reviewed (commented):
I was not feeling well these days, so I only went through the diff.

The evaluation tests and the build tests seem sound. As @philiptaron commits don't touch the actual build process, the building behavior should stay the same since I pushed my part of changes.

As long as the tests build, I approve @philiptaron's changes.

philiptaron committed:
buildEnv: explicitly set __structuredAttrs = false

builder.pl reads all configuration from %ENV, which is fundamentally
incompatible with __structuredAttrs = true (values go into a JSON file
instead). Explicitly opt out so buildEnv doesn't break when
structuredAttrsByDefault is enabled.

The override is placed in the final merge block so it cannot be
circumvented via derivationArgs.

Update the structuredAttrs test from asserting build failure to
verifying that the override works: the build succeeds even when
derivationArgs.__structuredAttrs = true, and eval tests confirm the
attribute is always false.

Suggested-by: Sietse Ringers <SFrijters>

MattSturgeon commented on pkgs/build-support/buildenv/default.nix:
```diff
@@ -121,6 +121,10 @@ lib.makeOverridable (
           (lib.remove null)
         ];
       in
+      assert
+        args ? name
+        || (args ? pname && args ? version)
+        || throw "buildEnv: expects arguments 'pname' and 'version', or 'name'";
```
For context, the assertion was originally added in 19f546749c67a741a9a5770ad93417a4d7116911

ShamrockLee replied:
> For context, the assertion was originally added in 19f546749c67a741a9a5770ad93417a4d7116911

That was a workaround I suggested for that PR to throw a meaningful error message when name attributes are missing.
